cpp-client/deephaven/dhclient/src/utility/table_maker.cc

```diff
@@ -59,11 +67,64 @@ TableHandle TableMaker::MakeTable(const TableHandleManager &manager) {
+  auto schema = MakeSchema();
+
+  // Extract the data_vec column.
+  auto data_vec = MakeReservedVector<std::shared_ptr<arrow::Array>>(column_infos_.size());
+  for (const auto &info : column_infos_) {
+    data_vec.push_back(info.data_);
+  }
+  auto data = GetColumnsNotEmpty();
+  return arrow::Table::Make(std::move(schema), std::move(data));
+}
+
+std::vector<std::shared_ptr<arrow::Array>> TableMaker::GetColumnsNotEmpty() const {
+  std::vector<std::shared_ptr<arrow::Array>> result;
+  for (const auto &info : column_infos_) {
+    result.emplace_back(info.data_);
+  }
+  if (result.empty()) {
+    throw std::runtime_error(DEEPHAVEN_LOCATION_STR("Can't make table with no columns"));
+  }
+  return result;
+}
+
+std::shared_ptr<arrow::Schema> TableMaker::MakeSchema() const {
+  arrow::SchemaBuilder sb;
+  for (const auto &info : column_infos_) {
+    auto kv_metadata = std::make_shared<arrow::KeyValueMetadata>();
+    OkOrThrow(DEEPHAVEN_LOCATION_EXPR(kv_metadata->Set("deephaven:type",
+        info.deepaven_server_type_name_)));
+
+    auto field = std::make_shared<arrow::Field>(info.name_, info.arrow_type_, true,
+        std::move(kv_metadata));
+    OkOrThrow(DEEPHAVEN_LOCATION_EXPR(sb.AddField(field)));
+  }
+
+  return ValueOrThrow(DEEPHAVEN_LOCATION_EXPR(sb.Finish()));
+}
+
+TableMaker::ColumnInfo::ColumnInfo(std::string name, std::shared_ptr<arrow::DataType> arrow_type,
+    std::string deepaven_server_type_name, std::shared_ptr<arrow::Array> data) :
+    name_(std::move(name)), arrow_type_(std::move(arrow_type)),
+    deepaven_server_type_name_(std::move(deepaven_server_type_name)), data_(std::move(data)) {}
+TableMaker::ColumnInfo::ColumnInfo(ColumnInfo &&other) noexcept = default;
+TableMaker::ColumnInfo::~ColumnInfo() = default;
+
 namespace internal {
-TypeConverter::TypeConverter(std::shared_ptr<arrow::DataType> data_type,
-    std::string deephaven_type, std::shared_ptr<arrow::Array> column) :
-    dataType_(std::move(data_type)), deephavenType_(std::move(deephaven_type)),
-    column_(std::move(column)) {}
-    TypeConverter::~TypeConverter() = default;
+const char DeephavenServerConstants::kBool[] = "java.lang.Boolean";
+const char DeephavenServerConstants::kChar16[] = "char";
+const char DeephavenServerConstants::kInt8[] = "byte";
+const char DeephavenServerConstants::kInt16[] = "short";
+const char DeephavenServerConstants::kInt32[] = "int";
+const char DeephavenServerConstants::kInt64[] = "long";
+const char DeephavenServerConstants::kFloat[] = "float";
+const char DeephavenServerConstants::kDouble[] = "double";
+const char DeephavenServerConstants::kString[] = "java.lang.String";
+const char DeephavenServerConstants::kDateTime[] = "java.time.ZonedDateTime";
+const char DeephavenServerConstants::kLocalDate[] = "java.time.LocalDate";
+const char DeephavenServerConstants::kLocalTime[] = "java.time.LocalTime";
+const char DeephavenServerConstants::kList[] = "what.goes.here";
```

Comment:
Is this intentional?  Reads like TODO intent...

Reply:
Thanks. This was a placeholder that I forgot about.

Background: this information is only used when we upload new tables to the server. Uploading tables with a "list" column type was something we weren't planning to support for now.

But in the process of moving/hiding this, I decided "it would just be easier" to fully support it. So now, uploading list types is supported and there is a new unit test for it in `new_table_test.cc`